### ISSUE TITLE
experiment with a Scala 3 version of Magnolia

### ISSUE THREAD

**Issue by propensive** [softwaremill/magnolia#246]
We would like to attempt to write a variant of Magnolia for Scala 3, which adheres to (as close as possible) the same API as Scala 2.x users have. There's likely to be a lot of experimentation involved here, as we learn Scala 3 and its quirks...

@kubukoz has kindly offered to work on this.

**Comment by joroKr21** [comment]
Just saw this: lampepfl/dotty#8811

Not saying we shouldn't go ahead with experimenting but it may turn out that macros will not be such a big problem for migration after all.

**Comment by neko-kai** [comment]
@joroKr21 
This isn't what it seems, this just allows you to compile and publish a Scala 2 macro using dotty (IFF you can manage to write a macro without using quasiquotes or `c.universe.reify` because these aren't supported)
You still have to write a Scala 3 macro to run on Scala 3, the purpose is to allow publishing a single cross-compiled jar with macros for 2 & 3, but it remains to be seen whether this ability is going to get used or most everyone will just publish different jars for different Scalas.

**Comment by joroKr21** [comment]
Ah I must have known it sounded too good to be true!
> (IFF you can manage to write a macro without using quasiquotes or c.universe.reify because these aren't supported)

That sounds like a severe limitation though.

**Comment by bishabosha** [comment]
> (IFF you can manage to write a macro without using quasiquotes or `c.universe.reify` because these aren't supported) - for consumption by Scala 2 compiler!

Hey, so actually the only thing you can write is `def foo = macro Macros.fooImpl`, where `Macros.fooImpl` has already been compiled by scala 2, so as long as you isolate your scala 2 macro implementations to a separate library then you can ship a scala 3 library that when read from scala 2.13 uses the correct macro implementation

**Comment by neko-kai** [comment]
@bishabosha Yeah, what i mean is that you can compile the Scala 2 macro implementation itself, a function `(c: Context): c.Tree`, with the Scala 3 compiler too, as long as you can avoid calling Scala 2 macros.

**Comment by bishabosha** [comment]
@neko-kai Oh yeah I see what you mean

**Comment by maphi** [comment]
After looking into the magnolia code i think that the solution via Scala 3 `Mirror` is not sufficient to support the full feature set of magnolia. It is lacking two things:
 - detect read only cases
 - details to fill the `Typename` class

Not sure though how to achieve that. I think there is something similar to `TypeTag` in Scala 3 which would probably solve the  second point but i think the first one would only be solvable via Scala 3 macros.

**Comment by propensive** [comment]
I think that some of those missing features shouldn't be too important. Detecting read-only cases is a relatively new feature anyway, and it's primarily a performance enhancement. Am I correct in saying that the `Typename`'s `owner` field should just be the full name of the package it's in? It would surprise me if we can't get this... And would anyone mind too much if we didn't support value classes in Scala 3?

**Comment by ghostdogpr** [comment]
Another thing that the typeclass derivation feature doesn't give you is annotations. I'm not sure if it's possible to get them without a macro?

**Comment by maphi** [comment]
Regarding annotations : It's true they are also missing. Never used them so i have no plan how to integrate them. 

Dropping features might be an option, or at least one could provide a subset like "magnolia core" offering the basic features in Scala 3.

 I've been also thinking about if providing derivation for value classes is at all a good thing. It should probably be at least something to make explicit and not beeing derived automatically as default. Not to mention that AFAIK value classes should not be used at all and can easily be replaced by opaque types in Scala 3.

@propensive AFAICT `owner` is the package name but i didn't look into the implementation. I agree shouldn't be too hard to get that. Much harder probably are the typearguments, but i wouldn't mind to drop that as well. The whole details in `Typename` are mostly usefull for implementing `Show` like instances which is only a single use case.

**Comment by maphi** [comment]
Also i think the most important thing to have is making `magnolia` compatible between Scala 2 and 3 so dropping features would require to drop them for Scala 2 too.

**Comment by maphi** [comment]
Apparently shapeless already got the annotations thingy: https://github.com/milessabin/shapeless/blob/shapeless-3/modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala

@ghostdogpr is that sufficient?

**Comment by ghostdogpr** [comment]
@maphi if that's accessible from inside the derivation code, I guess?

FYI, Caliban makes use of these Magnolia features (in addition to the basic features):
- `ReadOnlyCaseClass` but that's just for performance so it doesn't matter if we now rely on Scala 3 derivation
- `isValueClass` (so that the instance for the underlying type is used), but we can recommend using opaque types instead
- `isObject` (I think we can know that with Scala 3)
- `annotations`
- `typeName.full` and `typename.short`
- `typeArguments`

**Comment by maphi** [comment]
Hm i guess we can make the annotations available based on that code. I think the hardest part will be the `typeArguments`. What is the use case in caliban?

**Comment by maphi** [comment]
Regarding `isObject`: does it need to differentiate between case class X() and case object X ?

**Comment by ghostdogpr** [comment]
> Regarding `isObject`: does it need to differentiate between case class X() and case object X ?

It doesn't, so we could just check if there are no fields instead.

> Hm i guess we can make the annotations available based on that code. I think the hardest part will be the `typeArguments`. What is the use case in caliban?

We generate a GraphQL type from a Scala type, but since GraphQL doesn't have generics, we need to generate concrete types when generics are used in Scala. So let's say we have a case class `case class Box[A](a: A)`, if we use `Box[A]` it will generate a GraphQL type `BoxA` that contains an `A` and if we use `Box[B]` it will generate a GraphQL type `BoxB` that contains a `B`. Names must be unique.

**Comment by maphi** [comment]
Ok, makes sense. I'm currently diving into the Scala 3 macro stuff and it should be able to make that work AFAICT.

**Comment by joroKr21** [comment]
Only `Mirror` will not do but I think we can fill in the missing functionality with macros 👍

**Comment by maphi** [comment]
> Only `Mirror` will not do but I think we can fill in the missing functionality with macros 👍

Absolutely. I'm making some good progress with that right now.

**Comment by maphi** [comment]
@ghostdogpr @propensive 

### Regarding owner/type parameters

I was able to extract the owner and type parameters for a lot of types (see https://gist.github.com/maphi/3ef290599bca683a41bbffcbfa9d735b)

But what should be done for things like union or intersection types? Of course we could provide a sealed trait for all cases, but I think this is beyond the scope of this library and it would not cover other use cases. Instead this feature could be put into a module or another library and we should provide the quoting context instead, so that users have the flexibility to cover their use case. I'd also like to add that I'll provide help here if needed.

Regarding GraphQL:

I'm not sure how you would tackle e.g. union types in the GraphQL schema. Maybe one should explicitly provide a naming for that (e.g. via annotation) as the scheme might become difficult to read when names are derived automatically and must be unique.

### Reagarding annotations

I haven't tried yet but I saw annotation stuff while experimenting with the type parameter things. We also have the example how to extract annotations in shapeless, so I think this will be relatively easy to solve. 

### Conclusion

Before we start with an implementation the scope of magnolia should be discussed

**Comment by ghostdogpr** [comment]
Good question about union and intersection types. In Caliban I can definitely see the need to support union types and we could auto-generate the name, for example `A | B` would give a graphql union type called `AOrB` (similar to how we auto-name `List[A]`), and we have ways to override that name. I don't think we would need intersection types.

So I think it would be really cool if it was supported, but totally fine if it's left for future improvements.

**Comment by kubukoz** [comment]
paging @adamw for the greater good of tapir ;)

**Comment by joroKr21** [comment]
I would suggest to get sealed traits and enums working first and worry about union types later.

**Comment by maphi** [comment]
I've created a PR https://github.com/propensive/magnolia/pull/275. Some things are WIP or i don't know how to solve them yet or they would need some discussion which approach to take.

Missing/Unresolved:
- subtyping for sealed traits (wasn't able to get that to work yet, but should be doable)
- uses sbt instead of fury (I lack experience with fury, help needed)
- construct methods for product types missing
- for some annotations it is not clear to me where they would even be written (see TODO in PR)
- derive opaque types (not clear if this would be possible at all)
- derive value classes (not clear if this would be possible at all)
- get TypeInfo may not be exhaustive for all possible types yet (see code)

Changes:
- derivation is done via trait so that methods combine/dispatch are easily accessible
- rename gen to derived to match Scala 3 derivation
- Typename was replaced by TypeInfo

**Comment by heksesang** [comment]
Looks like the Scala 3 code is in the repo now – any plan to make a release soon, so people can start using it?

**Comment by propensive** [comment]
Yes, I'm planning to make the release as soon as possible, hopefully in the next couple of days. By the way, the Scala 2 version is still available on the `legacy` branch, and won't be going away any time soon.

**Comment by adamw** [comment]
Since we've got 2.0.0-Mx released, closing this. I'll create a separate issue to investigate union types.
